File: examples/pizzas_static_go/domain/pizza.go
package domain

import (
	"time"
	"github.com/google/uuid"
)

type Pizza struct {
	ID        string    `json:"id"`
	Name string `json:"name"`
	Description string `json:"description"`
	Toppings []Topping `json:"toppings"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func NewPizza(name string, description string, toppings []Topping) *Pizza {
	a := &Pizza{
		ID:        uuid.New().String(),
		Name: name,
		Description: description,
		Toppings: toppings,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *Pizza) Validate() error {
	if a.Name == "" {
		return &ValidationError{Field: "name", Message: "name can't be blank"}
	}
	if a.Description == "" {
		return &ValidationError{Field: "description", Message: "description can't be blank"}
	}
	return nil
}

File: examples/pizzas_domain/lib/pizzas_domain/pizza/pizza.rb
require 'hecks/mixins/model'

module PizzasDomain
  class Pizza
    autoload :Topping, "pizzas_domain/pizza/topping"

    include Hecks::Model

    attribute :name
    attribute :description
    attribute :toppings, default: [], freeze: true

    private

    def validate!
      raise ValidationError.new("name can't be blank", field: :name, rule: :presence) if name.nil? || (name.respond_to?(:empty?) && name.empty?)
      raise ValidationError.new("description can't be blank", field: :description, rule: :presence) if description.nil? || (description.respond_to?(:empty?) && description.empty?)
    end
  end
end
